api/v1alpha2/contentlibraryitem_types.go

```diff
@@ -0,0 +1,374 @@
+	// Type describes the type of controller.
+	Type VirtualControllerType `json:"type"`
+
+	// Key describes the controller's device key.
+	Key int32 `json:"key"`
+}
+
+// +kubebuilder:validation:Enum=FlatVer1;FlatVer2;SeSparse;SparseVer1;SparseVer2
+
+// VirtualDiskType defines the allowed types of virtual disks.
+type VirtualDiskType string
+
+const (
+	VirtualDiskTypeFlatVer1   VirtualDiskType = "FlatVer1"
+	VirtualDiskTypeFlatVer2   VirtualDiskType = "FlatVer2"
+	VirtualDiskTypeSeSparse   VirtualDiskType = "SeSparse"
+	VirtualDiskTypeSparseVer1 VirtualDiskType = "SparseVer1"
+	VirtualDiskTypeSparseVer2 VirtualDiskType = "SparseVer2"
+)
+
+type VirtualDiskInfo struct {
+	// Key describes the disk's device key.
+	Key int32 `json:"key"`
+
+	// ControllerKey describes the device key of the controller to which the
+	// disk is connected.
+	ControllerKey int32 `json:"controllerKey"`
+
+	// +optional
+
+	// Capacity describes the capacity of the disk.
+	Capacity resource.Quantity `json:"capacity,omitempty"`
+
+	// Type describes the type of virtual disk.
+	Type VirtualDiskType `json:"type"`
+}
+
+// FileInfo represents the information of a file in a content library item in
+// vCenter.
+type FileInfo struct {
+	// Name describes the name of the file.
+	Name string `json:"name"`
+
+	// +optional
+
+	// SizeInBytes describes the total number of bytes used by the file on disk.
+	SizeInBytes resource.Quantity `json:"sizeInBytes,omitempty"`
+
+	// +optional
+
+	// Version describes the version of the library item file.
+	// This value is incremented when the file is modified on disk.
+	Version string `json:"version,omitempty"`
+
+	// +optional
+
+	// Cached describes whether or not the file is available locally on disk.
+	Cached bool `json:"cached,omitempty"`
+
+	// +optional
+
+	// StorageURI describes the fully-qualified path to the file on a datastore,
+	// ex. "[my-datastore-1] library-1/item-1/file1.vmdk".
+	// This URL is useful for creating a device that is backed by this file
+	// (i.e. mounting an ISO file via a virtual CD-ROM device).
+	StorageURI string `json:"storageURI,omitempty"`
+
+	// +optional
+
+	// DiskInfo describes information about the file if it is a virtual disk,
+	// otherwise this field is omitted.
+	DiskInfo *VirtualDiskInfo `json:"diskInfo,omitempty"`
+}
+
+// ContentLibraryItemSpec defines the desired state of a ContentLibraryItem.
+type ContentLibraryItemSpec struct {
+	// ID describes the unique identifier used to find the library item in
+	// vCenter.
+	//
+	// Please note this value depends on spec.libraryType:
+	// - LibraryType=ContentLibrary -- ID is a content library item UUID.
+	// - LibraryType=Inventory      -- ID is a vSphere VM managed object ID.
+	ID string `json:"id"`
+
+	// +optional
+	// +kubebuilder:default=ContentLibrary
+
+	// LibraryType describes the type of library to which this item belongs.
+	//
+	// Defaults to ContentLibrary.
+	LibraryType LibraryType `json:"libraryType,omitempty"`
```

Comment:
should it inherit from the parent library type or can a library have mixed types of item, i.e. vm + iso, or can this type be different from the parent library type ?

Reply:
How do you mean? This is the library type, not the type of content.

Reply:
Since an item belongs to a library, the library that the item belongs to has a library type, now this item spec allows specifying the library type, can it be different from the type of the parent library? in the inventory library type case, how will the library of this item be identified? in the contentlibrary type case, each item is unique, so it doesn't need a library to identify.

Reply:
So the purpose of this is to understand the `ID` field of the library item:

```go
// ContentLibraryItemSpec defines the desired state of a ContentLibraryItem.
type ContentLibraryItemSpec struct {
	// ID describes the unique identifier used to find the library item in
	// vCenter.
	//
	// Please note this value depends on spec.libraryType:
	// - LibraryType=ContentLibrary -- ID is a content library item UUID.
	// - LibraryType=Inventory      -- ID is a vSphere VM managed object ID.
	ID string `json:"id"`
```

Maybe instead of being a string, `ID` itself should be a struct that has a value and type? It's the same intent no matter how you slice it though.

Reply:
Resolved offline.